packages/router/src/page-loader.js

```diff
@@ -0,0 +1,48 @@
+import { useState } from 'react'
+
+import { createNamedContext } from './internal'
+
+const PageLoadingContext = createNamedContext('PageLoading')
+
+const PageLoader = ({ spec, delay, params }) => {
+  const [cache, setCache] = useState({})
+  const [pageName, setPageName] = useState(null)
+  const [loading, setLoading] = useState(false)
+
+  const { loader, name } = spec
+
+  let loadingTimeout
+
+  const loadedPage = cache[name]
+  if (loading) {
+    // noop
+  } else if (loadedPage) {
+    if (pageName != loadedPage.name) {
+      setPageName(loadedPage.name)
+    }
+  } else {
+    loadingTimeout = setTimeout(() => setLoading(true), delay)
+    loader().then((module) => {
+      cache[name] = module.default
+      setCache(cache)
+      setPageName(name)
+      setLoading(false)
+      if (loadingTimeout) {
+        clearTimeout(loadingTimeout)
+      }
+    })
+  }
```

Comment:
This is called on every render, but maybe we should be explicit about when we want it to be called by wrapping it in useEffect.